internal/cephfs/groupcontrollerserver.go

```diff
@@ -0,0 +1,756 @@
+// false it will call createSnapshot and get the snapshot details for the
+// volume and add the snapshotID and volumeID to the volume group journal omap.
+// If any error occurs other than ErrInProgress it will delete the snapshots
+// and undo the reservation and return the error.
+func (cs *ControllerServer) createSnapshotAddToVolumeGroupJournal(
+	ctx context.Context,
+	req *csi.CreateVolumeGroupSnapshotRequest,
+	freeze bool,
+	vgo *store.VolumeGroupOptions,
+	vgs *store.VolumeGroupSnapshotIdentifier,
+	cr *util.Credentials,
+	fsMap map[string]core.FSQuiesceClient,
+	secrets map[string]string) (
+	*[]csi.CreateSnapshotResponse,
+	error,
+) {
+	var err error
+	var resp *csi.CreateSnapshotResponse
+
+	defer func() {
+		if err != nil && !errors.Is(err, cerrors.ErrQuiesceInProgress) {
+			err = cs.deleteSnapshotsAndUndoReservation(ctx, vgs, cr, fsMap, secrets)
+			if err != nil {
+				log.ErrorLog(ctx, "failed to delete snapshot and undo reservation: %v", err)
+			}
+		}
+	}()
+
+	responses := make([]csi.CreateSnapshotResponse, 0)
+	for _, volID := range req.GetSourceVolumeIds() {
+		if !freeze {
+			err = fsQuiesceWithExpireTimeout(ctx, vgo.RequestName, fsMap)
+			if err != nil {
+				log.ErrorLog(ctx, "failed to quiesce filesystem with timeout: %v", err)
+
+				return nil, err
+			}
+		}
+		// Create the snapshot for the volumeID
+		clusterID := getClusterIDForVolumeID(fsMap, volID)
+		if clusterID == "" {
+			return nil, fmt.Errorf("failed to get clusterID for volumeID %s", volID)
+		}
+		req := &csi.CreateSnapshotRequest{
+			SourceVolumeId: volID,
+			Name:           vgs.FsVolumeGroupSnapshotName + "-" + volID,
+			Secrets:        secrets,
+			Parameters: map[string]string{
+				"clusterID": clusterID,
+			},
+		}
+
+		resp, err = cs.createSnapshotAndAddMapping(ctx, req, vgo, vgs)
+		if err != nil {
+			// Handle cleanup
+			log.ErrorLog(ctx, "failed to create snapshot: %v", err)
+
+			return nil, err
+		}
+		responses = append(responses, *resp)
+	}
+
+	if !freeze {
+		err = releaseFSQuiesce(ctx, vgo.RequestName, fsMap)
+		if err != nil {
+			log.ErrorLog(ctx, "failed to release filesystem quiesce: %v", err)
+
+			return nil, err
+		}
+	}
+
+	return &responses, nil
+}
+
+// releaseSubvolumeQuiesce releases the quiesce of the subvolumes and subvolume
+// groups in the filesystems for the volumeID's present in the
+// CreateVolumeGroupSnapshotRequest.
+func releaseFSQuiesce(ctx context.Context,
+	requestName string,
+	fsMap map[string]core.FSQuiesceClient,
+) error {
+	inProgress := false
+	var err error
+	var data *admin.QuiesceInfo
+	for _, fm := range fsMap {
+		// UnFreeze the filesystems, subvolumes and subvolume groups
+		data, err = fm.ReleaseFSQuiesce(ctx, requestName)
+		if err != nil {
+			log.ErrorLog(ctx, "failed to release filesystem quiesce: %v", err)
+
+			return err
+		}
+		state := core.GetQuiesceState(data.State)
+		if state != core.Released {
+			inProgress = true
+
+			continue
```

Comment:
This continue is not required ? 
or is some parts of code missing ?

Reply:
it can be removed not required